### ISSUE TITLE
Stream Analytics

### ISSUE THREAD

**Issue by ronoc** [seiflotfy/chronos#16]
(In talks with Neil)

"A way to handle stream analytics. Considering our current experience, if we had a way to define generic analytics that could be handled in memory as a pre insert process would be super cool".

**Comment by njpatel** [comment]
https://github.com/snowplow/snowplow

**Comment by seiflotfy** [comment]
Yeah I was thinking of that over night (in my sleep). Its the direction we
ought to take. E.g: We need to be able to solve the "distinct value
problem" via pre_insert. This could be don't by pushing to redis or keeping
stuff in our own memory and dumping a status (the bitmap) every x
insertions or y minutes to disk. Then load those again when loading up.
Another issue I think needs to be handled in memory is "live rolling
retention". Also think I have a way for it.
This direction takes us away from being an ES based project which is
totally worth it.

Also after taking a look last night at snowplow they also follow the
pattern of "we define what an event is".
http://snowplowanalytics.com/analytics/event-dictionaries-and-data-models/event-dictionaries-and-schemas.html
I
